experimental/sql/src/projection_pushdown.py

```diff
@@ -0,0 +1,138 @@
+    else:
+      return self.flatten([self.find_cols_in_expr(r.op) for r in op.regions])
+
+  def find_cols_upstream(self, op: RelAlg.Operator) -> list[str]:
+    """
+    Returns the list of cols that are used in this operation or its parents. If
+    the list contains None, a ModuleOp (so the output) was hit, showing that all
+    columns need to be kept.
+    """
+    if isinstance(op, ModuleOp):
+      return [None]
+    if isinstance(op, RelAlg.Select):
+      return self.flatten([
+          self.find_cols_in_expr(o) for o in op.predicates.ops
+      ]) + self.find_cols_upstream(op.parent_op())
+    if isinstance(op, RelAlg.Project):
+      return self.flatten(
+          [self.find_cols_in_expr(o) for o in op.projections.ops])
+    if isinstance(op, RelAlg.Aggregate):
+      return [s.data for s in op.col_names.data]
+
+  def find_schema(self, op: RelAlg.Operator) -> list[str]:
+    """
+    Returns the list of cols that are the input schema of this operation.
+    """
+    if isinstance(op, RelAlg.Table):
+      return [o.elt_name.data for o in op.schema.ops]
+    if isinstance(op, RelAlg.Select):
+      return self.find_schema(op.input.op)
+    if isinstance(op, RelAlg.Project):
+      return [s.data for s in op.names.data]
+    if isinstance(op, RelAlg.Aggregate):
+      return [s.data for s in op.col_names.data]
+
+
+@dataclass
+class ProjectionInference(ProjectionOptimizer):
+
+  # This rewriter infers columns in front of all Operators (but projections and
+  # tables) that map away all columns that aren't needed by this Operator or
+  # further up the tree.
+
+  @op_type_rewrite_pattern
+  def match_and_rewrite(self, op: RelAlg.Operator, rewriter: PatternRewriter):
+    if isinstance(op, RelAlg.Project) or isinstance(
+        op, RelAlg.Table) or isinstance(op, RelAlg.SchemaElement):
+      return
+    cols = list(dict.fromkeys(self.find_cols_upstream(op)))
+    if cols.__contains__(None):
+      return
+
+    new_proj = RelAlg.Project.get(
+        rewriter.move_region_contents_to_new_regions(op.input),
+        Region.from_operation_list([RelAlg.Column.get(s) for s in cols]),
+        ArrayAttr.from_list([StringAttr.from_str(s) for s in cols]))
+
+    # TODO: I mainly need this pattern match to call the proper builder. This
+    # could maybe be done nicer.
+    if isinstance(op, RelAlg.Select):
+      rewriter.replace_matched_op(
+          RelAlg.Select.get(
+              Region.from_operation_list([new_proj]),
+              rewriter.move_region_contents_to_new_regions(op.predicates)))
+    elif isinstance(op, RelAlg.Aggregate):
+      rewriter.replace_matched_op(
+          RelAlg.Aggregate.get(Region.from_operation_list([new_proj]),
+                               [s.data for s in op.col_names.data],
+                               [s.data for s in op.functions.data],
+                               [s.data for s in op.res_names.data]))
+
+
+@dataclass
+class IdentityProjectionRemover(ProjectionOptimizer):
```

Comment:
If I didn't forget about a corner case, then this will certainly go into the thesis. I think this rewriter is a good example of something that is very easy on this abstraction level, but close to impossible on e.g. the iterators level.